### ISSUE TITLE
Unable to use RE-direct URL on spotify web

### ISSUE THREAD

**Issue by erikd234** [brim-borium/spotify_sdk#84]
Hi I am new to flutter web development and I cant get Spotify to authenticate on the web. Right now I have set my redirect URL to my http://localhost:63389 where my flutter app is running after calling 
SpotifySdk.connectToSpotifyRemote()

I get to this page here 
![image](https://user-images.githubusercontent.com/76464314/112404894-6f191b00-8cce-11eb-8a2f-6a6bd321bb83.png)
when I proceed after giving permissions it redirects me to my local host.
I now have two local hosts open and neither is connecting to spotify.
I can't play any songs or do anything and in my console in VScode it still tells me that its "Connecting to Spotify..."

My question is how can I authenticate my flutter application on the web? More directly how does the redirect URI work? It works perfectly on iOS.

**Comment by rish07** [comment]
Hey, @erikd234 have you whitelisted this redirect URL in the Spotify dev console?

**Comment by erikd234** [comment]
> Hey, @erikd234 have you whitelisted this redirect URL in the Spotify dev console?

Yes I have. i have double checked this as well.

**Comment by erikd234** [comment]
@itsMatoosh Hello Matoosh,  i think my problem is I dont understand what a redirecct url does. What should I put as a redirect url if my local application is running on  http://localhost:63389 ? What even is a redirect url?

**Comment by fotiDim** [comment]
Closing this as this is a support question. It is mostly suited for Stack Overflow. Nevertheless to get you unstuck the redirect url is used by Spotify (needs to be registered under your dev console) in order to redirect back to your app after successful authentication. I suggest to read the official Spotify documentation for this. In your case you can set `http://localhost:63389/` at the redirect url. It is fine for local development.

**Comment by itsMatoosh** [comment]
@erikd234 It is important that the redirect URL you specify points to a page with HTML contents as specified in the readme. 
When Spotify authenticates the user, it will redirect them to the redirect URL. The script on that page will then communicate with the spotify_sdk plugin to complete the authentication.

**Comment by erikd234** [comment]
Thanks for getting back to me! I had the script in the read me on a live
website that I own and not on my local host. To specify For I was running
the flutter application on localhost and I made the redirect url a live
website with the read me HTML page. Would that be why it isnt working? Do i
need to make sure that both the html contents in the readme and my web app
are on the same domain?

**Comment by erikd234** [comment]
@itsMatoosh Now after I authenticate my app and the redirect url works properly I am getting this error in flutter V 2.1.0-13.0.pre.175.   
![image](https://user-images.githubusercontent.com/76464314/112883510-3d63d380-9083-11eb-8973-b523d5601677.png)
To reproduce this error just run the flutter example in debug mode, hot restart once otherwise the connect to Spotify button wont work (due to the web startup bug that's currently an issue). When I hit the connect to Spotify button on the top left of the web app, I get redirected to authenticate on Spotify I sign in and it does that properly, then I go to the redirect url to my webpage which has the script from the read me on it. Then in my debug console I get the error above. Ideas?

**Comment by erikd234** [comment]
Further update, the redirect URL must be working because I successfully obtain an authenticationToken whenever I run SpotifySdk.getAuthenticationToken I keep getting the same error above "connectToSpotify failed with: TypeError: Cannot read property 'Player' of undefined". I even tried passing the the authentication token as a parameter to connectToSpotifyRemote(). Still no luck? Any ideas and should I just open a new issue?

![image](https://user-images.githubusercontent.com/76464314/112905229-ef5cc900-909e-11eb-95c8-3dddb6d2e564.png)

**Comment by erikd234** [comment]
![image](https://user-images.githubusercontent.com/76464314/112906227-b291d180-90a0-11eb-973b-bbf7d2cb4da3.png)
Could this issue have something to do with the dart interacting with JavaScript @JS("Spotify.Player").

**Comment by itsMatoosh** [comment]
@erikd234 I believe the error you're getting is related to the app running in debug mode. As you mentioned, there is currently an error when initializing the player in debug mode. 

Whenever the Spotify Web SDK loads correctly, it calls the "onSpotifyWebPlaybackSDKReady" callback, which in turn sets up the Spotify player with this plugin. Because of the debug mode error, this callback never runs.

There is a check in the web implementation constructor that looks at whether the "onSpotifyWebPlaybackSDKReady" callback was already set to a method by the plugin, and if it was then it skips initializing the Spotify player to avoid duplicate Spotify initialization. When you hot-reload you basically short-circuit this check, as the "onSpotifyWebPlaybackSDKReady" is already defined but the Spotify player is still not correctly initialized. That's why you're able to press the "Connect" button, but still get an error later on.

So basically it's still the same error with debug mode which I think needs to be solved on Flutter's side.

**Comment by itsMatoosh** [comment]
@erikd234 Regarding the redirect URL, I believe it needs to be hosted on the same domain as your app because of the CORS policy of the browser.
